README
LogFetcher
==========
Version: 0.9.0
Date: 2018-01-24

Use SSH to fetch log files from a configurable list of machines and store them locally
takeing into account that the logfiles on the remote hosts are rotated

Enjoy!

Tobias Oetiker <tobi@oetiker.ch>


CHANGES
0.10.3 2018-01-24  more debugging output

0.10.2 2018-01-23  made hostchanel keepalive happy even when files get filtered

0.10.1 2016-09-30  made file filter more robust

0.10.0 2016-09-30  added minAge option

0.8.2 2016-01-22  improve handling of host channel loss

0.0.0 2015-12-10  Tobias Oetiker (tobi@oetiker.ch)

- started project


lib/LogFetcher.pm
package LogFetcher;

use Mojo::Base 'Mojolicious';
use LogFetcher::Config;
our $VERSION = "0.1.0";

=head1 NAME

AzeOP - the application class

=head1 SYNOPSIS

 use Mojolicious::Commands;
 Mojolicious::Commands->start_app('LogFetcher');

=head1 DESCRIPTION

Configure the mojolicious engine to run our application logic

=cut

=head1 ATTRIBUTES

LogFetcher has all the attributes of L<Mojolicious> plus:

=cut

=head2 config

use our own plugin directory and our own configuration file:

=cut

has config => sub {
    my $app = shift;
    LogFetcher::Config->new(
        app => $app,
        file => $ENV{LOGFETCHER_CFG} || $app->home->rel_file('etc/logfetcher.cfg' )
    );
};


sub startup {
    my $app = shift;
    @{$app->commands->namespaces} = (__PACKAGE__.'::Command');
    $app->commands->message("Usage:\n\n".$app->commands->extract_usage."\nCommands:\n\n");
}

1;

=head1 COPYRIGHT

Copyright (c) 2015 by OETIKER+PARTNER AG. All rights reserved.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=cut


lib/LogFetcher/Config.pm
package LogFetcher::Config;

use Mojo::Base -base;
use Mojo::JSON qw(decode_json);
use Mojo::Util qw(slurp);
use Carp qw(croak);
use Mojo::Exception;
use Data::Processor;
use Data::Processor::ValidatorFactory;

=head1 NAME

LogFetcher::Config - the Config access  class

=head1 SYNOPSIS

 use LogFetcher::Config;
 my $conf = LogFetcher::Config->new(app=>$app,file=>'config.json');
 my $hash = $conf->cfgHash;
 print $hash->{GENERAL}{value};

=head1 DESCRIPTION

Load and preprocess a configuration file in json format.

=head1 ATTRIBUTES

All the attributes from L<Mojo::Base> as well as:

=head2 app

pointing to the app

=cut

has 'app';

has validatorFactory => sub {
    Data::Processor::ValidatorFactory->new;
};

has validator => sub {
    Data::Processor->new(shift->schema);
};

=head2 file

the path of the config file

=cut

has 'file';


=head2 SCHEMA

the flattened content of the config file

=cut




use Data::Dumper;

my $CONSTANT_RE = '[_A-Z]+';

has schema => sub {
    my $self = shift;
    my $vf = $self->validatorFactory;
    my $string = $vf->rx('^.*$','expected a string');
    my $integer = $vf->rx('^\d+$','expected an integer');

    return {
        GENERAL => {
            description => 'general settings',
            members => {
                logFile => {
                    validator => $vf->file('>>','writing'),
                    description => 'absolute path to log file',
                },
                logLevel => {
                    validator => $vf->rx('(?:debug|info|warn|error|fatal)','Pick a logLevel of debug, info, warn, error or fatal'),
                    description => 'mojo log level - debug, info, warn, error or fatal'
                },
                logCheckInterval => {
                    description => 'log check interval in seconds',
                    validator => $integer,
                },
                statusLogInterval => {
                    description => 'how often to report the log sync status in seconds',
                    validator => $integer,
                },
                transferTaskLimit => {
                    optional => 1,
                    description => 'how many transfer tasks can be in progress in parallel',
                    validator => $integer,
                },
                timeout => {
                    description => 'how long to wait for tasks to start',
                    validator => $integer,
                    default => 5,
                },
            },
        },
        CONSTANTS => {
            description => 'define constants fo be used in globPattern and destinationFile properties.',
            optional => 1,
            members => {
                $CONSTANT_RE => {
                    regex => 1,
                    description => 'value of the constant',
                    validator => $string,
                }
            }

        },
        HOSTS => {
            description => 'where does our data come from.',
            array => 1,
            members => {
                name => {
                    description => 'identifier for this host for logfiles',
                    validator => $string,
                },
                statType => {
                    optional => 1,
                    description => 'pick gnu or bsd for the type of stat comand on the remote system',
                    validator => $vf->rx('^(gnu|bsd)$','expected gnu or bsd'),
                    default => 'gnu',
                },
                sshConnect => {
                    description => 'ssh arguments',
                    array => 1,
                    validator => $string
                },
                logFiles => {
                    description => 'a map of globs on the remote machine',
                    array => 1,
                    members => {
                        globPattern => {
                            description => 'a glob pattern to find all rotated logfile versions on the remote hosts. you can use ${CONSTANTS}',
                            validator => $string
                        },
                        filterRegexp => {
                            optional => 1,
                            description => <<'DOC_END',
A regular expression to filter the files found by the globPattern. If the RE contains captures, they
can be accessed as C<${RXMATCH_1}>, C<${RXMATCH_2}> and so on ... (maximal 5 captures).
DOC_END
                            validator => $string
                        },
                        minAge => {
                            optional => 1,
                            description => <<'DOC_END',
The minimum age of the file in seconds to be considered for fetching.  Use
this if you want to fetch files that get a new name daily.  Set the age to
one day, as this will cause the file to be fetched once it has not been
written to for at least a day.
DOC_END
                            validator => $integer
                        },
                        destinationFile => {
                            description => 'where to store the file you can use ${CONSTANTS} and strftime formatting',
                            validator => $string
                        }
                    }
                }
            }
        },
    };
};

=head2 rawCfg

raw config

=cut

has rawCfg => sub {
    my $self = shift;
    $self->n3kCommon->loadJSONCfg($self->file);
};


=head2 cfgHash

access the config hash

=cut

has cfgHash => sub {
    my $self = shift;
    my $cfg = $self->loadJSONCfg($self->file);
    my $validator = $self->validator;
    my $hasErrors;
    my $err = $validator->validate($cfg);
    for ($err->as_array){
        warn "$_\n";
        $hasErrors = 1;
    }
    die "Can't continue with config errors\n" if $hasErrors;
    # we need to set this real early to catch all the info in the logfile.
    $self->app->log->path($cfg->{GENERAL}{logFile});
    $self->app->log->level($cfg->{GENERAL}{logLevel});
    if (my $const = $cfg->{CONSTANTS}){
        my $CONST_MATCH = join('|',keys %$const);
        for my $host (@{$cfg->{HOSTS}}){
            for my $logFile (@{$host->{logFiles}}){
                for my $key (qw(globPattern destinationFile)){
                    $logFile->{$key} =~ s/\$\{($CONST_MATCH)\}/$const->{$1}/g;
                }
            }
        }
    }
    return $cfg;
};

=head1 METHODS

All the methods of L<Mojo::Base> as well as:

=head2 loadJSONCfg(file)

Load the given config, sending error messages to stdout and igonring /// lines as comments

=cut

sub loadJSONCfg {
    my $self = shift;
    my $file = shift;
    my $json = slurp($file);
    $json =~ s{^\s*//.*}{}gm;
    my $raw_cfg = eval { decode_json($json) };
    if ($@){
        if ($@ =~ /(.+?) at line (\d+), offset (\d+)/){
            my $warning = $1;
            my $line = $2;
            my $offset = $3;
            open my $json, '<', $file;
            my $c =0;
            warn "Reading ".$file."\n";
            warn "$warning\n\n";
            while (<$json>){
                chomp;
                $c++;
                if ($c == $line){
                    warn ">-".('-' x $offset).'.'."  line $line\n";
                    warn "  $_\n";
                    warn ">-".('-' x $offset).'^'."\n";
                }
                elsif ($c+3 > $line and $c-3 < $line){
                    warn "  $_\n";
                }
            }
            warn "\n";
            exit 1;
        }
        else {
            Mojo::Exception->throw("Reading ".$file.': '.$@);
        }
    }
    return $raw_cfg;
}


1;

__END__

=head1 COPYRIGHT

Copyright (c) 2015 by OETIKER+PARTNER AG. All rights reserved.

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=head1 HISTORY

 2014-12-16 to 0.0 first version

=cut

# Emacs Configuration
#
# Local Variables:
# mode: cperl
# eval: (cperl-set-style "PerlStyle")
# mode: flyspell
# mode: flyspell-prog
# End:
#
# vi: sw=4 et


lib/LogFetcher/HostChannel.pm
package LogFetcher::HostChannel;
use Mojo::Base -base;
use Mojo::IOLoop::ReadWriteFork;
use POSIX qw(strftime);
use File::Path qw(make_path);
use Time::HiRes qw(gettimeofday);
use Scalar::Util;


=head1 NAME

LogFetcher::HostChannel - host channel

=head1 SYNOPSIS

    my $cc = LogFetcher::HostChannel->new(sshArgs=>[ssh arguments],logFiles=>[log files]);
    $cc->fetch();

=head1 DESCRIPTION

Open an ssh connection with the given arguments. Then figure out the modification times for all
matching files on the remote system and map them to their local storage location. If the destination file
already exists, skip. Otherwhise launch a separate ssh process to download the file and store it locally.

=cut

=head1 ATTRIBUTES

=head2 sshConnect

argument array to pass on to ssh to open a connection to a remote host

=cut

has sshConnect => sub {
    return [];
};

=head2 statType

is the remote system providing gnu or bsd stat

=cut

has statType => sub {
    return 'gnu';
};

=head2 logFiles

a list of logfile hashes as defined in the config format. These are the files we are going to watch.

=cut

has logFiles => sub {
    return [];
};

=head2 name

identifier for this channel

=cut

has 'name';

=head2 log

a L<Mojo::Log> object

=cut

has 'log';

=head2 hostChannel

a L<Mojo::IOLoop::ReadWriteFork> instance connected to the remote host

=cut

has 'hostChannel';

=head2 statistics

a hash of statistics

=cut

=head2 gCfg

GENERAL section form the config file

=cut

has 'gCfg';

has stats => sub {
    return {
        filesTransfered => 0,
        bytesTransfered => 0,
        filesChecked => 0,
    };
};

my @defaultSshOpts = qw(-T -x -y -o BatchMode=yes -o ConnectTimeout=10);

=head1 METHODS

=head2 fetch

go through all the globs on the remote host and see which ones we already go here and
fetch the rest.

=cut

sub makePath {
    my $self = shift;
    my $working = shift;
    if ($working =~ m{^(/.+/)} and not -d $1){
        make_path($1,{error => \my $err});
        if (scalar @$err) {
            for my $diag (@$err) {
                my ($file, $message) = %$diag;
                if ($file eq '') {
                    $self->log->error($self->name.": mkdir $working: $message");
                }
                else {
                    $self->log->error($self->name.": problem creating $file: $message");
                }
            }
            return;
        }
    }
}

sub checkTimeStamp {
    my $self = shift;
    my $src = shift;
    my $timeStamp = shift;
    my $endCheck = shift;
    my $abort = shift;
    # if the fork handle goes out of scope, the fork gets destoryed ...
    # by saving it into this hash we keep it alive until it is done with its work
    state %rc;
    my $checkFork = Mojo::IOLoop::ReadWriteFork->new;
    $rc{$checkFork} = $checkFork;
    my $remoteTimeStamp = 99;
    my $read = '';
    my $firstRead;
    $checkFork->on(read => sub {
        my $checkFork = shift;
        my $chunk = shift;
        $read .= $chunk;
        $firstRead //= substr($chunk,1,256);
        if ($read =~ /<(\d+)>/){
            $remoteTimeStamp = $1;
        }
    });

    my $timeout = Mojo::IOLoop->timer($self->gCfg->{timeout} => sub {
        $checkFork->kill(9);
        $abort->("stamp check $src: TIMEOUT");
    });

    $checkFork->on(close => sub {
        Mojo::IOLoop->remove($timeout);
        my $checkFork = shift;
        my $exitValue = shift;
        my $signal = shift;
        delete $rc{$checkFork};
#        if (rand() > 0.5){
#            $abort->("timestamp check $src: random abort");
#            return;
#        }
        if ($exitValue != 0 or $signal){
            $abort->("timestamp check $src: SSH problem Signal $signal, ExitValue $exitValue: ".($firstRead//'no data'));
            return;
        }
        if ($remoteTimeStamp == $timeStamp){
            $self->log->debug($self->name.": timestamp check $src: verification ok");
            $endCheck->();
        }
        else {
            $abort->("timestamp check $src: timestamp is $remoteTimeStamp and not $timeStamp as expected.");
            return;
        }
    });
    $checkFork->on(error => sub {
        Mojo::IOLoop->remove($timeout);
        my $checkFork = shift;
        my $error = shift;
        delete $rc{$checkFork};
        $checkFork->kill(9);
        $abort->("stamp check $src: $error");
    });
    my $cmd = $self->statType eq 'gnu' ? "stat --format='<%Y>' $src" : "stat -f '<%Dm>' $src" ;
    my @sshArgs = (@{$self->sshConnect},@defaultSshOpts,$cmd);
    $self->log->debug($self->name.': ssh '.join(' ',@sshArgs));
    $checkFork->start(
        program => 'ssh',
        program_args => \@sshArgs,
    );

};

sub checkSumWorking {
    my $self = shift;
    my $working = shift;
    my $endTransfer = shift;
    my $abort = shift;
    # if the fork handle goes out of scope, the fork gets destoryed ...
    # by saving it into this hash we keep it alive until it is done with its work
    state %rc;
    my $checkFork = Mojo::IOLoop::ReadWriteFork->new;
    # save a copy
    $rc{$checkFork} = $checkFork;
    my $timeout = Mojo::IOLoop->timer(600 => sub {
        $checkFork->kill(9);
        $abort->("checksuming check $working: TIMEOUT");
    });

    $checkFork->on(close => sub {
        Mojo::IOLoop->remove($timeout);
        my $checkFork = shift;
        my $exitValue = shift;
        my $signal = shift;
        if ($exitValue == 0){
            $self->log->debug($self->name.": gunzip checksum $working - OK");
        }
        else {
            $abort->("checksum check for $working failed Signal: $signal, ExitValue $exitValue");
        }
        $endTransfer->();
        delete $rc{$checkFork};
    });
    $checkFork->on(error => sub {
        my $checkFork = shift;
        my $error = shift;
        Mojo::IOLoop->remove($timeout);
        $checkFork->kill(9);
        delete $rc{$checkFork};
        $abort->("checksum check for $working: $error");
    });
    my @gunzipArgs = (qw(--test --quiet),$working);
    $self->log->debug($self->name.': gunzip checksum test '.join(' ',@gunzipArgs));
    $checkFork->start(
        program => 'gunzip',
        program_args => \@gunzipArgs,
    );
};

# track the active transfers
my %doneFiles;
my %workFiles;

sub transferFile {
    my $self = shift;
    my $src = shift;
    my $dest = shift;
    my $timeStamp = shift;
    # if the transferFork handle goes out of scope, the fork gets destoryed ...
    # by saving it into this hash we keep it alive until it is done with its work
    state %rc;
    my $transferStarted = 0;
    my $working = $dest.'.working';
    $self->makePath($working);
    my $outWrite;
    my $taskLimit = $self->gCfg->{transferTaskLimit};
    if ($taskLimit and $taskLimit < scalar keys %workFiles){
        $self->log->info($self->name.": Not transfering $src yet since ".(scalar keys %workFiles)." transfers are in progress");
        return undef;
    };
    if ($workFiles{$working}){
        $self->log->warn($self->name.": fetch skipped $src $working transfer already in progress.");
        return undef;
    }
    if (not open($outWrite, '>', $working)){
        $self->log->warn($self->name.": fetch $src $working failed: $!.");
        return undef;
    }
    $workFiles{$working} = 1 ;
    my $transferFork = Mojo::IOLoop::ReadWriteFork->new;
    # stringify outside the callback to not
    # create a closure
    my $forkKey = "$transferFork";
    # the keep alive copy
    $rc{$forkKey} = $transferFork;
    my $delay = Mojo::IOLoop->delay(sub {
        my $delay = shift;
        if (my $error = $delay->data('error')){
            $self->log->error($self->name.": fetch $src $dest - $error");
            unlink $working;
        }
        else {
            rename $working,$dest;
            $doneFiles{$dest} = 1;
            $self->log->info($self->name.": fetch completed $src $dest ".$delay->data('perfData'));
            $self->stats->{filesTransfered}++;
        }
        # now it can go ... bye bye
        delete $workFiles{$working};
        delete $rc{$forkKey};
    });

    my $endTransfer = $delay->begin();
    my $endCheck = $delay->begin();
    Scalar::Util::weaken(my $wtf = $transferFork);
    my $abort = sub {
        my $error = shift;
        $delay->data('error',$error);
        $wtf->kill(9) if $wtf;
        $endTransfer->();
        $endCheck->();
     };

    my $timeoutHandler = sub {
        $abort->("TIMEOUT");
    };

    my $timeoutId = Mojo::IOLoop->timer($self->gCfg->{timeout} => $timeoutHandler);
    my $startTime;
    my $totalSize = 0;
    my $firstRead;
    $transferFork->on(read => sub {
        my $transferFork = shift;
        my $chunk = shift;
        $self->log->debug($self->name.": got ".length($chunk)." bytes on transferFork for $src");    
        $firstRead //= substr($chunk,1,256);
        if (not $transferStarted){
            $startTime = gettimeofday();
            # $self->log->debug($self->name.": fetch $src $dest first byte");
            $transferStarted = 1;
            $self->checkTimeStamp($src,$timeStamp,$endCheck,$abort);
        }
        Mojo::IOLoop->remove($timeoutId);
        $timeoutId = Mojo::IOLoop->timer($self->gCfg->{timeout} => $timeoutHandler);
        my $size = length($chunk);
        $self->stats->{bytesTransfered} += $size;
        $totalSize += $size;
        syswrite $outWrite,$chunk;
    });
    $transferFork->on(close => sub {
        my $transferFork = shift;
        my $exitValue = shift;
        my $signal = shift;
        Mojo::IOLoop->remove($timeoutId);
        if ($signal or $exitValue or not $transferStarted){
            if (not $delay->data('error')){
                if ($signal){
                    $delay->data('error',"aborted: Signal $signal");
                }
                else {
                    $delay->data('error',"failed: ExitValue $exitValue: ".($firstRead//'no error info'));
                }
            }
            $endTransfer->();
            $endCheck->();
        }
        else {
            $delay->data('perfData',sprintf("%.1f MB @ %.1f MB/s",
                $totalSize/1024/1024,$totalSize/1024/1024/(gettimeofday()-$startTime)));
            $self->checkSumWorking($working,$endTransfer,$abort);
        }
    });
    $transferFork->on(error => sub {
        my $transferFork = shift;
        my $error = shift;
        $delay->data('error',"fetch $src $dest failed: $error");
        Mojo::IOLoop->remove($timeoutId);
        $endTransfer->();
        $endCheck->();
    });
    # no use double compressing things ...
    my $cmd = ($src =~ m/\.gz$/ ? 'cat ' : 'gzip -c ').$src;
    my @sshArgs = (@{$self->sshConnect},@defaultSshOpts,$cmd);
    $self->log->info($self->name.": fetch started $src -> $working.");
    $self->log->debug($self->name.': ssh '.join(' ',@sshArgs));
    $transferFork->start(
        program => 'ssh',
        program_args => \@sshArgs,
        conduit => 'pipe',
    );
};

# open a new fork
has 'lastLogInfoLine';
has 'hostChannelFirstRead';
has 'hostChannel';

sub makeHostChannel {
    my $self = shift;
    my $controlFork = Mojo::IOLoop::ReadWriteFork->new;
    my $read = '';
    my $firstRead;

    $self->hostChannelFirstRead('no data');
    $self->lastLogInfoLine(time);
    $controlFork->on(read => sub {
        my $controlFork = shift;
        my $chunk = shift;
        $read .= $chunk;
        $self->lastLogInfoLine(time);
        $self->log->debug($self->name.": got ".length($chunk)." bytes on hostChannel");
        while ( $read =~ s/^(.*?)<LOG_FILE><(\d+)><(\d+)><(.+?)><NL>//s ){
            if (not $firstRead and $1){
                $firstRead = $1;
                $self->hostChannelFirstRead($firstRead);
            }
            my ($id,$time,$file) = ($2,$3,$4);
            if (my $minAge = $self->logFiles->[$id]{minAge}){
                if (time - $time <= $minAge){
                    $self->log->debug($self->name.": skipping $file as is not $minAge seconds old");
                    next;
                }
            }
            my %match = (
                RXMATCH_1 => '',
                RXMATCH_2 => '',
                RXMATCH_3 => '',
                RXMATCH_4 => '',
                RXMATCH_5 => '',
            );
            if (my $filter = $self->logFiles->[$id]{filterRegexp}){
                if ($file !~ $filter){
                    $self->log->debug($self->name.": skipping $file as it does not match filterRexexp /$filter/");
                    next;
                }
                %match = (
                    RXMATCH_1 => $1 // '',
                    RXMATCH_2 => $2 // '',
                    RXMATCH_3 => $3 // '',
                    RXMATCH_4 => $4 // '',
                    RXMATCH_5 => $5 // '',
                );
            }
            my $dest = strftime($self->logFiles->[$id]{destinationFile},localtime($time));
            $dest =~ s/\$\{(RXMATCH_[1-5])\}/$match{$1}/g;
            $self->stats->{filesChecked}++;
            if (not $doneFiles{$dest} and not -f $dest){
                $self->transferFile($file,$dest,$time);
            }
        }
        if (not $firstRead and $read and length($read) > 10){
            $firstRead = substr($read,0,320);
            chomp($firstRead);
            $firstRead =~ s/[\n\r]+/\\n/g;
            $self->hostChannelFirstRead($firstRead);
        }
        if ( $read =~ s/<ERROR>//){
            my $error = $read;
            $error =~ s/[\n\r]+/\\n/g;
            $self->log->error($self->name.": Host Channel Problem: ".($error//'no error info'));
        }
    });

    $controlFork->on(close => sub {
        my $controlFork = shift;
        my $exitValue = shift;
        my $signal = shift;
        if ($exitValue != 0 and not $signal){
            $self->log->error($self->name.": Host Channel SSH Problem ExitValue $exitValue: ".($firstRead//'no error info'));
        }
        else {
            $self->log->error($self->name.": Host Channel Closed: Signal $signal");
        }
        $self->hostChannel(undef);
    });

    $controlFork->on(error => sub {
        my $controlFork = shift;
        my $error = shift;
        $self->log->error($self->name.': Host Channel Closed - '.$error);
        $self->hostChannel(undef);
    });

    $self->log->debug($self->name.': ssh '.join(' ',@{$self->sshConnect}).' (hostChannel)');
    $controlFork->start(
        program => 'ssh',
        program_args => [@{$self->sshConnect},@defaultSshOpts],
        conduit => 'pipe'
    );
    return $controlFork;
};

sub fetch {
    my $self = shift;
    if (not $self->hostChannel){
        my $hc = $self->makeHostChannel;
        if ($hc){
            $self->hostChannel($hc);
        }
        else {
            return;
        }
    }
    if (time - $self->lastLogInfoLine > $self->gCfg->{timeout}+$self->gCfg->{logCheckInterval}){
        $self->log->error($self->name.': hostChannel not reacting anymore ... lets get a new one. ('.$self->hostChannelFirstRead.')');
        $self->hostChannel->kill(9);
        return;
    };
    my $logFiles = $self->logFiles;
    for (my $id = 0;$id < scalar @$logFiles;$id++){
        $self->hostChannel->write(
            ($self->statType eq 'gnu'
              ? "stat --format='<LOG_FILE><$id><%Y><%n><NL>' "
              : "stat -f '<LOG_FILE><$id><%Dm><%SN><NL>' ")
            . $logFiles->[$id]{globPattern}
            . " || echo '<ERROR>'\n"
        );
    }
}

1;
__END__

=back

=head1 COPYRIGHT

Copyright (c) 2015 by OETIKER+PARTNER AG. All rights reserved.

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=head1 HISTORY

 2011-05-30 to 1.0 first version

=cut

# Emacs Configuration
#
# Local Variables:
# mode: cperl
# eval: (cperl-set-style "PerlStyle")
# mode: flyspell
# mode: flyspell-prog
# End:
#
# vi: sw=4 et


lib/LogFetcher/Command/fetch.pm
package LogFetcher::Command::fetch;
use Mojo::Base 'Mojolicious::Command';
use Getopt::Long 2.25 qw(:config posix_default no_ignore_case);
use LogFetcher::HostChannel;

=head1 NAME

LogFetcher::Command::fetch - Log File Syncer

=head1 SYNOPSIS

 ./logfetcher.pl fetch

=head1 DESCRIPTION

Run the logfetcher

=cut

has description => <<'EOF';
copy rotating logfiles from a remote machine to a local archive
EOF

has usage => <<"EOF";
usage: $0 fetch [OPTIONS]

These options are available:

  --verbose   talk about your work

EOF

my %opt;

has log => sub { shift->app->log };
has cfg => sub { shift->app->config->cfgHash };


sub run {
    my $self   = shift;
    local @ARGV = @_ if scalar @_;
    GetOptions(\%opt, 'verbose|v');
    if ($opt{verbose}){
        $ENV{MOJO_LOG_LEVEL} = 'debug';
        $self->log->level('debug');
        $self->log->handle(\*STDOUT);
    }
    my $app = $self->app;


    # setup our tasks
    for my $host (@{$self->cfg->{HOSTS}}){
        # since the $channel is used in the recurring callbacks
        # it does not get DESTORYED prematurely (or every actyally)
        # but here this is fine
        my $channel = LogFetcher::HostChannel->new(%$host,log=>$self->log,gCfg=>$self->cfg->{GENERAL});
        Mojo::IOLoop->recurring( $self->cfg->{GENERAL}{logCheckInterval} => sub {
            $channel->fetch;
        });
        Mojo::IOLoop->recurring( $self->cfg->{GENERAL}{statusLogInterval} => sub {
            my $s = $channel->stats;
            $self->log->info($channel->name.": "
                . "$s->{filesChecked} files checked, "
                . "$s->{filesTransfered} files transfered, "
                . "$s->{bytesTransfered} bytes transfered"
            );
            for (qw(filesChecked filesTransfered bytesTransfered)) {
                $s->{$_} = 0;
            }
        });
        $channel->fetch();
    }

    Mojo::IOLoop->start;
    return $self;
}

1;
__END__

=back

=head1 COPYRIGHT

Copyright (c) 2015 by OETIKER+PARTNER AG. All rights reserved.

=head1 LICENSE

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

=head1 AUTHOR

S<Tobias Oetiker E<lt>tobi@oetiker.chE<gt>>

=head1 HISTORY

 2015-05-30 to 1.0 first version

=cut

# Emacs Configuration
#
# Local Variables:
# mode: cperl
# eval: (cperl-set-style "PerlStyle")
# mode: flyspell
# mode: flyspell-prog
# End:
#
# vi: sw=4 et
